File: LeetCode/2/2.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* curr = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode* l3 = curr;

    int carry = 0;
    while (l1 != NULL || l2 != NULL) {
        int sum = 0;

        if (l1 != NULL) {
            sum += l1->val;
            l1 = l1->next;
        }
        if (l2 != NULL) {
            sum += l2->val;
            l2 = l2->next;
        }

        sum += carry;
        carry = sum / 10;
        sum = sum % 10;

        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr->next->val = sum;
        curr->next->next = NULL;
        curr = curr->next;
    }

    if (carry != 0) {
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr->next->val = carry;
        curr->next->next = NULL;
    }

    return l3->next;
}

File: LeetCode/2/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        curr = head
        carry = 0
        while l1 != None or l2 != None or carry != 0:
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            totalSum = x + y + carry
            carry = totalSum // 10
            curr.next = ListNode(totalSum % 10)
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return head.next
        